File: benchmark/pystone.cpp
#include "cpp_benchmarks.h"

#include <array>
#include <string>

namespace benchmark_cpp
{
    namespace
    {
        constexpr int64_t ident1 = 1;
        constexpr int64_t ident2 = 2;
        constexpr int64_t ident3 = 3;
        constexpr int64_t ident4 = 4;
        constexpr int64_t ident5 = 5;

        struct PystoneRecord
        {
            PystoneRecord *ptr_comp = nullptr;
            int64_t discr = 0;
            int64_t enum_comp = 0;
            int64_t int_comp = 0;
            const char *string_comp = nullptr;
        };

        struct PystoneContext
        {
            int64_t int_glob = 0;
            bool bool_glob = false;
            char char1_glob = '\0';
            char char2_glob = '\0';
            std::array<int64_t, 51> array1_glob{};
            std::array<std::array<int64_t, 51>, 51> array2_glob{};
            PystoneRecord ptr_glb;
            PystoneRecord ptr_glb_next;
            PystoneRecord proc1_next_record;
        };

        PystoneRecord pystone_copy_record(const PystoneRecord &record)
        {
            return record;
        }

        bool pystone_func3(int64_t enum_par_in)
        {
            int64_t enum_loc = enum_par_in;
            return enum_loc == ident3;
        }

        int64_t pystone_func1(char char_par1, char char_par2)
        {
            char char_loc1 = char_par1;
            char char_loc2 = char_loc1;
            if(char_loc2 != char_par2)
            {
                return ident1;
            }
            return ident2;
        }

        bool pystone_func2(const std::string &str_par_i1,
                           const std::string &str_par_i2)
        {
            int64_t int_loc = 1;
            char char_loc = '\0';
            while(int_loc <= 1)
            {
                if(pystone_func1(str_par_i1[int_loc],
                                 str_par_i2[int_loc + 1]) == ident1)
                {
                    char_loc = 'A';
                    int_loc = int_loc + 1;
                }
            }
            if(char_loc >= 'W' && char_loc <= 'Z')
            {
                int_loc = 7;
            }
            if(char_loc == 'X')
            {
                return true;
            }
            if(str_par_i1 > str_par_i2)
            {
                int_loc = int_loc + 7;
                (void)int_loc;
                return true;
            }
            return false;
        }

        int64_t pystone_proc6(const PystoneContext &ctx, int64_t enum_par_in)
        {
            int64_t enum_par_out = enum_par_in;
            if(!pystone_func3(enum_par_in))
            {
                enum_par_out = ident4;
            }
            if(enum_par_in == ident1)
            {
                enum_par_out = ident1;
            }
            else if(enum_par_in == ident2)
            {
                if(ctx.int_glob > 100)
                {
                    enum_par_out = ident1;
                }
                else
                {
                    enum_par_out = ident4;
                }
            }
            else if(enum_par_in == ident3)
            {
                enum_par_out = ident2;
            }
            else if(enum_par_in == ident4)
            {
            }
            else if(enum_par_in == ident5)
            {
                enum_par_out = ident3;
            }
            return enum_par_out;
        }

        int64_t pystone_proc7(int64_t int_par_i1, int64_t int_par_i2)
        {
            int64_t int_loc = int_par_i1 + 2;
            int64_t int_par_out = int_par_i2 + int_loc;
            return int_par_out;
        }

        PystoneRecord *pystone_proc3(PystoneContext &ctx,
                                     PystoneRecord *ptr_par_out)
        {
            if(ctx.ptr_glb.ptr_comp != nullptr)
            {
                ptr_par_out = ctx.ptr_glb.ptr_comp;
            }
            else
            {
                ctx.int_glob = 100;
            }
            ctx.ptr_glb.int_comp = pystone_proc7(10, ctx.int_glob);
            return ptr_par_out;
        }

        PystoneRecord *pystone_proc1(PystoneContext &ctx,
                                     PystoneRecord *ptr_par_in)
        {
            ctx.proc1_next_record = pystone_copy_record(ctx.ptr_glb);
            PystoneRecord &next_record = ctx.proc1_next_record;
            ptr_par_in->ptr_comp = &ctx.proc1_next_record;
            ptr_par_in->int_comp = 5;
            next_record.int_comp = ptr_par_in->int_comp;
            next_record.ptr_comp = ptr_par_in->ptr_comp;
            next_record.ptr_comp = pystone_proc3(ctx, next_record.ptr_comp);
            if(next_record.discr == ident1)
            {
                next_record.int_comp = 6;
                next_record.enum_comp =
                    pystone_proc6(ctx, ptr_par_in->enum_comp);
                next_record.ptr_comp = ctx.ptr_glb.ptr_comp;
                next_record.int_comp = pystone_proc7(next_record.int_comp, 10);
            }
            else
            {
                *ptr_par_in = pystone_copy_record(next_record);
            }
            next_record.ptr_comp = nullptr;
            return ptr_par_in;
        }

        int64_t pystone_proc2(const PystoneContext &ctx, int64_t int_par_io)
        {
            int64_t int_loc = int_par_io + 10;
            int64_t enum_loc = 0;
            while(true)
            {
                if(ctx.char1_glob == 'A')
                {
                    int_loc = int_loc - 1;
                    int_par_io = int_loc - ctx.int_glob;
                    enum_loc = ident1;
                }
                if(enum_loc == ident1)
                {
                    break;
                }
            }
            return int_par_io;
        }

        void pystone_proc4(PystoneContext &ctx)
        {
            bool bool_loc = ctx.char1_glob == 'A';
            bool_loc = bool_loc || ctx.bool_glob;
            (void)bool_loc;
            ctx.char2_glob = 'B';
        }

        void pystone_proc5(PystoneContext &ctx)
        {
            ctx.char1_glob = 'A';
            ctx.bool_glob = false;
        }

        void pystone_proc8(PystoneContext &ctx,
                           std::array<int64_t, 51> &array1_par,
                           std::array<std::array<int64_t, 51>, 51> &array2_par,
                           int64_t int_par_i1, int64_t int_par_i2)
        {
            int64_t int_loc = int_par_i1 + 5;
            array1_par[int_loc] = int_par_i2;
            array1_par[int_loc + 1] = array1_par[int_loc];
            array1_par[int_loc + 30] = int_loc;
            for(int64_t int_index = int_loc; int_index < int_loc + 2;
                ++int_index)
            {
                array2_par[int_loc][int_index] = int_loc;
            }
            array2_par[int_loc][int_loc - 1] =
                array2_par[int_loc][int_loc - 1] + 1;
            array2_par[int_loc + 20][int_loc] = array1_par[int_loc];
            ctx.int_glob = 5;
        }

        int64_t pystone_checksum(const PystoneContext &ctx)
        {
            int64_t result = ctx.int_glob;
            if(ctx.bool_glob)
            {
                result += 1;
            }
            if(ctx.char1_glob == 'A')
            {
                result += 10;
            }
            if(ctx.char2_glob == 'B')
            {
                result += 100;
            }
            result += ctx.array1_glob[8];
            result += ctx.array1_glob[9];
            result += ctx.array1_glob[38];
            result += ctx.array2_glob[8][7];
            result += ctx.array2_glob[8][8];
            result += ctx.array2_glob[28][8];
            result += ctx.ptr_glb.discr;
            result += ctx.ptr_glb.enum_comp;
            result += ctx.ptr_glb.int_comp;
            result += ctx.ptr_glb.ptr_comp->int_comp;
            return result;
        }
    }  // namespace

    int64_t pystone_run(int64_t n)
    {
        PystoneContext ctx;
        ctx.ptr_glb_next = {nullptr, 0, 0, 0, nullptr};
        ctx.ptr_glb = {&ctx.ptr_glb_next, ident1, ident3, 40,
                       "DHRYSTONE PROGRAM, SOME STRING"};
        std::string string1_loc = "DHRYSTONE PROGRAM, 1'ST STRING";
        ctx.array2_glob[8][7] = 10;

        for(int64_t i = 0; i < n; ++i)
        {
            pystone_proc5(ctx);
            pystone_proc4(ctx);
            int64_t int_loc1 = 2;
            int64_t int_loc2 = 3;
            std::string string2_loc = "DHRYSTONE PROGRAM, 2'ND STRING";
            int64_t enum_loc = ident2;
            ctx.bool_glob = !pystone_func2(string1_loc, string2_loc);
            int64_t int_loc3 = 0;
            while(int_loc1 < int_loc2)
            {
                int_loc3 = 5 * int_loc1 - int_loc2;
                int_loc3 = pystone_proc7(int_loc1, int_loc2);
                int_loc1 = int_loc1 + 1;
            }
            pystone_proc8(ctx, ctx.array1_glob, ctx.array2_glob, int_loc1,
                          int_loc3);
            ctx.ptr_glb = *pystone_proc1(ctx, &ctx.ptr_glb);
            char char_index = 'A';
            while(char_index <= ctx.char2_glob)
            {
                if(enum_loc == pystone_func1(char_index, 'C'))
                {
                    enum_loc = pystone_proc6(ctx, ident1);
                }
                char_index = static_cast<char>(char_index + 1);
            }
            int_loc3 = int_loc2 * int_loc1;
            double int_loc2_float =
                static_cast<double>(int_loc3) / static_cast<double>(int_loc1);
            int_loc2_float = 7 * (int_loc3 - int_loc2_float) - int_loc1;
            (void)int_loc2_float;
            int_loc1 = pystone_proc2(ctx, int_loc1);
            (void)int_loc1;
        }

        return pystone_checksum(ctx);
    }

    int64_t pystone_items(int64_t n) { return n; }
}  // namespace benchmark_cpp

File: benchmark/pystone.py
# Adapted from the Computer Language Benchmarks Game pystone.python3.
# Source: https://github.com/dundee/pybenchmarks/blob/master/bencher/programs/pystone/pystone.python3
#
# The Computer Language Benchmarks Game
# Copyright 2008-2012 Isaac Gouy
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#    Redistributions of source code must retain the above copyright notice, this
#    list of conditions and the following disclaimer.
#
#    Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
#    Neither the name of "The Computer Language Benchmarks Game" nor the name of
#    "The Computer Language Shootout Benchmarks" nor the name "bencher" nor the
#    names of its contributors may be used to endorse or promote products
#    derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# CloverVM adaptation notes:
# - the benchmark harness externally times run(n), so pystone's internal timing
#   and null-loop calibration are removed;
# - setup-only unsupported constructs are rewritten: list multiplication, list
#   comprehensions, and unpacking assignment;
# - Proc0 returns a deterministic integer checksum for harness validation.

Ident1 = 1
Ident2 = 2
Ident3 = 3
Ident4 = 4
Ident5 = 5

TRUE = 1
FALSE = 0


class Record:
    def __init__(
        self, PtrComp=None, Discr=0, EnumComp=0, IntComp=0, StringComp=0
    ):
        self.PtrComp = PtrComp
        self.Discr = Discr
        self.EnumComp = EnumComp
        self.IntComp = IntComp
        self.StringComp = StringComp

    def copy(self):
        return Record(
            self.PtrComp,
            self.Discr,
            self.EnumComp,
            self.IntComp,
            self.StringComp,
        )


def make_zero_array():
    result = []
    for i in range(51):
        result.append(0)
    return result


def make_zero_matrix():
    result = []
    for i in range(51):
        result.append(make_zero_array())
    return result


IntGlob = 0
BoolGlob = FALSE
Char1Glob = "\0"
Char2Glob = "\0"
Array1Glob = None
Array2Glob = None
PtrGlb = None
PtrGlbNext = None


def run(n):
    return Proc0(n)


def Proc0(loops):
    global IntGlob
    global BoolGlob
    global Char1Glob
    global Char2Glob
    global Array1Glob
    global Array2Glob
    global PtrGlb
    global PtrGlbNext

    IntGlob = 0
    BoolGlob = FALSE
    Char1Glob = "\0"
    Char2Glob = "\0"
    Array1Glob = make_zero_array()
    Array2Glob = make_zero_matrix()

    PtrGlbNext = Record()
    PtrGlb = Record()
    PtrGlb.PtrComp = PtrGlbNext
    PtrGlb.Discr = Ident1
    PtrGlb.EnumComp = Ident3
    PtrGlb.IntComp = 40
    PtrGlb.StringComp = "DHRYSTONE PROGRAM, SOME STRING"
    String1Loc = "DHRYSTONE PROGRAM, 1'ST STRING"
    Array2Glob[8][7] = 10

    for i in range(loops):
        Proc5()
        Proc4()
        IntLoc1 = 2
        IntLoc2 = 3
        String2Loc = "DHRYSTONE PROGRAM, 2'ND STRING"
        EnumLoc = Ident2
        BoolGlob = not Func2(String1Loc, String2Loc)
        while IntLoc1 < IntLoc2:
            IntLoc3 = 5 * IntLoc1 - IntLoc2
            IntLoc3 = Proc7(IntLoc1, IntLoc2)
            IntLoc1 = IntLoc1 + 1
        Proc8(Array1Glob, Array2Glob, IntLoc1, IntLoc3)
        PtrGlb = Proc1(PtrGlb)
        CharIndex = "A"
        while CharIndex <= Char2Glob:
            if EnumLoc == Func1(CharIndex, "C"):
                EnumLoc = Proc6(Ident1)
            CharIndex = chr(ord(CharIndex) + 1)
        IntLoc3 = IntLoc2 * IntLoc1
        IntLoc2 = IntLoc3 / IntLoc1
        IntLoc2 = 7 * (IntLoc3 - IntLoc2) - IntLoc1
        IntLoc1 = Proc2(IntLoc1)

    result = IntGlob
    if BoolGlob:
        result = result + 1
    if Char1Glob == "A":
        result = result + 10
    if Char2Glob == "B":
        result = result + 100
    result = result + Array1Glob[8]
    result = result + Array1Glob[9]
    result = result + Array1Glob[38]
    result = result + Array2Glob[8][7]
    result = result + Array2Glob[8][8]
    result = result + Array2Glob[28][8]
    result = result + PtrGlb.Discr
    result = result + PtrGlb.EnumComp
    result = result + PtrGlb.IntComp
    result = result + PtrGlb.PtrComp.IntComp
    PtrGlb = None
    PtrGlbNext = None
    Array1Glob = None
    Array2Glob = None
    return result


def Proc1(PtrParIn):
    NextRecord = PtrGlb.copy()
    PtrParIn.PtrComp = NextRecord
    PtrParIn.IntComp = 5
    NextRecord.IntComp = PtrParIn.IntComp
    NextRecord.PtrComp = PtrParIn.PtrComp
    NextRecord.PtrComp = Proc3(NextRecord.PtrComp)
    if NextRecord.Discr == Ident1:
        NextRecord.IntComp = 6
        NextRecord.EnumComp = Proc6(PtrParIn.EnumComp)
        NextRecord.PtrComp = PtrGlb.PtrComp
        NextRecord.IntComp = Proc7(NextRecord.IntComp, 10)
    else:
        PtrParIn = NextRecord.copy()
    NextRecord.PtrComp = None
    return PtrParIn


def Proc2(IntParIO):
    IntLoc = IntParIO + 10
    while 1:
        if Char1Glob == "A":
            IntLoc = IntLoc - 1
            IntParIO = IntLoc - IntGlob
            EnumLoc = Ident1
        if EnumLoc == Ident1:
            break
    return IntParIO


def Proc3(PtrParOut):
    global IntGlob

    if PtrGlb is not None:
        PtrParOut = PtrGlb.PtrComp
    else:
        IntGlob = 100
    PtrGlb.IntComp = Proc7(10, IntGlob)
    return PtrParOut


def Proc4():
    global Char2Glob

    BoolLoc = Char1Glob == "A"
    BoolLoc = BoolLoc or BoolGlob
    Char2Glob = "B"


def Proc5():
    global Char1Glob
    global BoolGlob

    Char1Glob = "A"
    BoolGlob = FALSE


def Proc6(EnumParIn):
    EnumParOut = EnumParIn
    if not Func3(EnumParIn):
        EnumParOut = Ident4
    if EnumParIn == Ident1:
        EnumParOut = Ident1
    elif EnumParIn == Ident2:
        if IntGlob > 100:
            EnumParOut = Ident1
        else:
            EnumParOut = Ident4
    elif EnumParIn == Ident3:
        EnumParOut = Ident2
    elif EnumParIn == Ident4:
        pass
    elif EnumParIn == Ident5:
        EnumParOut = Ident3
    return EnumParOut


def Proc7(IntParI1, IntParI2):
    IntLoc = IntParI1 + 2
    IntParOut = IntParI2 + IntLoc
    return IntParOut


def Proc8(Array1Par, Array2Par, IntParI1, IntParI2):
    global IntGlob

    IntLoc = IntParI1 + 5
    Array1Par[IntLoc] = IntParI2
    Array1Par[IntLoc + 1] = Array1Par[IntLoc]
    Array1Par[IntLoc + 30] = IntLoc
    for IntIndex in range(IntLoc, IntLoc + 2):
        Array2Par[IntLoc][IntIndex] = IntLoc
    Array2Par[IntLoc][IntLoc - 1] = Array2Par[IntLoc][IntLoc - 1] + 1
    Array2Par[IntLoc + 20][IntLoc] = Array1Par[IntLoc]
    IntGlob = 5


def Func1(CharPar1, CharPar2):
    CharLoc1 = CharPar1
    CharLoc2 = CharLoc1
    if CharLoc2 != CharPar2:
        return Ident1
    else:
        return Ident2


def Func2(StrParI1, StrParI2):
    IntLoc = 1
    while IntLoc <= 1:
        if Func1(StrParI1[IntLoc], StrParI2[IntLoc + 1]) == Ident1:
            CharLoc = "A"
            IntLoc = IntLoc + 1
    if CharLoc >= "W" and CharLoc <= "Z":
        IntLoc = 7
    if CharLoc == "X":
        return TRUE
    else:
        if StrParI1 > StrParI2:
            IntLoc = IntLoc + 7
            return TRUE
        else:
            return FALSE


def Func3(EnumParIn):
    EnumLoc = EnumParIn
    if EnumLoc == Ident3:
        return TRUE
    return FALSE
